base/broadcast.jl

```diff
@@ -973,21 +952,23 @@ end
     tuplebroadcast(longest_tuple(nothing, bc.args), bc)
 @inline tuplebroadcast(::NTuple{N,Any}, bc) where {N} =
     ntuple(k -> _broadcast_getindex(bc, k), Val(N))
-longest_tuple(::Nothing, t::TupleLL{<:Tuple})   = longest_tuple(t.head, t.rest)
-longest_tuple(::Nothing, t::TupleLL)            = longest_tuple(nothing, t.rest)
-longest_tuple(l::Tuple, t::TupleLL{<:Tuple})    = longest_tuple(_longest_tuple(l, t.head), t.rest)
-longest_tuple(l::Tuple, t::TupleLL)             = longest_tuple(l, t.rest)
-longest_tuple(l::Tuple, t::TupleLL{TupleLLEnd}) = l
-longest_tuple(l::Tuple, ::TupleLLEnd)           = l
-longest_tuple(::Nothing, t::TupleLL{<:Broadcasted,TupleLLEnd}) = longest_tuple(nothing, t.head.args)
-longest_tuple(::Nothing, t::TupleLL{<:Broadcasted}) = longest_tuple(longest_tuple(nothing, t.head.args), t.rest)
-longest_tuple(l::Tuple, t::TupleLL{<:Broadcasted,TupleLLEnd}) = longest_tuple(l, t.head.args)
-longest_tuple(l::Tuple, t::TupleLL{<:Broadcasted}) = longest_tuple(longest_tuple(l, t.head.args), t.rest)
+# This is a little tricky: find the longest tuple (first arg) within the list of arguments (second arg)
+# Start with nothing as a placeholder and go until we find the first tuple in the argument list
+longest_tuple(::Nothing, t::Tuple{Tuple,Vararg{Any}}) = longest_tuple(t[1], tail(t))
+# Or recurse through nested broadcast expressions
+longest_tuple(::Nothing, t::Tuple{Broadcasted,Vararg{Any}}) = longest_tuple(longest_tuple(nothing, t[1].args), tail(t))
+longest_tuple(::Nothing, t::Tuple) = longest_tuple(nothing, tail(t))
+# And then compare it against all other tuples we find in the argument list or nested broadcasts
+longest_tuple(l::Tuple, t::Tuple{Tuple,Vararg{Any}}) = longest_tuple(_longest_tuple(l, t[1]), tail(t))
+longest_tuple(l::Tuple, t::Tuple) = longest_tuple(l, tail(t))
+longest_tuple(l::Tuple, ::Tuple{}) = l
+longest_tuple(l::Tuple, t::Tuple{Broadcasted}) = longest_tuple(l, t[1].args)
+longest_tuple(l::Tuple, t::Tuple{Broadcasted,Vararg{Any}}) = longest_tuple(longest_tuple(l, t[1].args), tail(t))
```

Comment:
I think we can simplify this logic. One option might be consider everything pairwise:
```julia
get_longest_tuple(t::Tuple) = t
get_longest_tuple(bc::Broadcasted) = get_longest_tuple(bc.args...)
function get_longest_tuple(A::Tuple, B::Tuple)
    if length(B) === 1 || length(A) === length(B)
        return A
    elseif length(A) === 1
        return B
    else
        throw(DimensionMismatch("tuples $A and $B could not be broadcast to a common size"))
    end
end
get_longest_tuple(A, B) = get_longest_tuple(get_longest_tuple(A), get_longest_tuple(B))
get_longest_tuple(A, B, t...) = get_longest_tuple(get_longest_tuple(A, B), get_longest_tuple(t...))
```
This drops the functionality of allowing non-Tuples in head position.

I think here's an even simpler version, which separates out even more the computation of the `_broadcast_tuplelen` and checking that the result shape is valid (pairwise):
```julia
_broadcast_len(t::NTuple{N, Any}) where {N} = N
_broadcast_len(bc::Broadcasted) = _broadcast_tuplelen(bc.args...)
# _broadcast_len(::Any) = 1 # are we supposed to assume that anything else isn't broadcast?
_broadcast_len(l, t...) = _broadcast_check_len(_broadcast_tuplelen(A), _broadcast_tuplelen(t...))
function _broadcast_check_len(A::Int, B::Int)
    if B == 1 || A == B
        return A
    elseif A == 1
        return B
    else
        throw(DimensionMismatch("tuples of length $A and $B could not be broadcast to a common size"))
    end
end
```

This would require that we delete the `tuplebroadcast` function though:
```julia
 @inline copy(bc::Broadcasted{Style{Tuple}}) =
      ntuple(k -> _broadcast_getindex(bc, k), Val(_broadcast_len(bc)))
```